src/decoder.rs

```diff
@@ -81,43 +82,48 @@ impl TransactionState {
                         self.setup = Some(
                             SetupFields::from_data_packet(packet));
                         // Wait for ACK.
-                        DecodeStatus::Continue
+                        Continue
                     },
-                    _ => DecodeStatus::Invalid
+                    _ => Invalid
                 }
             }
             // ACK then completes the transaction.
-            (Some(SETUP), Some(DATA0), ACK) => DecodeStatus::Done,
+            (Some(SETUP), Some(DATA0), ACK) => Done,
 
             // IN may be followed by NAK or STALL, completing transaction.
-            (_, Some(IN), NAK | STALL) => DecodeStatus::Done,
+            (_, Some(IN), NAK | STALL) => Done,
             // IN or OUT may be followed by DATA0 or DATA1, wait for status.
             (_, Some(IN | OUT), DATA0 | DATA1) => {
                 if packet.len() >= 3 {
                     let range = 1 .. (packet.len() - 2);
                     self.payload = packet[range].to_vec();
-                    DecodeStatus::Continue
+                    Continue
                 } else {
-                    DecodeStatus::Invalid
+                    Invalid
                 }
             },
             // An ACK or NYET then completes the transaction.
-            (Some(IN | OUT), Some(DATA0 | DATA1), ACK | NYET) => DecodeStatus::Done,
+            (Some(IN | OUT), Some(DATA0 | DATA1), ACK | NYET) => Done,
             // OUT may also be completed by NAK or STALL.
-            (Some(OUT), Some(DATA0 | DATA1), NAK | STALL) => DecodeStatus::Done,
+            (Some(OUT), Some(DATA0 | DATA1), NAK | STALL) => Done,
+
+            // PING may be followed by ACK, NAK or STALL.
+            (Some(PING), Some(PING), ACK | NAK | STALL) => Done,
 
             // Any other case is not a valid part of a transaction.
-            _ => DecodeStatus::Invalid,
+            _ => Invalid,
         })
     }
 
     fn completed(&self) -> bool {
         use PID::*;
         // A transaction is completed if it has 3 valid packets and is
-        // acknowledged with an ACK or NYET handshake.
-        match (self.count, self.last) {
-            (3, Some(ACK | NYET)) => true,
-            (..)                  => false
+        // acknowledged with an ACK or NYET handshake, except for PING
+        // transactions which complete after 2 with an ACK or NAK.
+        match (self.first, self.count, self.last) {
+            (Some(PING), 2, Some(ACK | NAK) ) => true,
```

Comment:
I guess this should match `STALL` too?

Reply:
No, I don't think it should. This method is used in `transfer_status` to decide whether the transfer should progress to the next state. For other transactions that happens on an ACK or NYET, which both indicate success. For PING, I accepted both ACK and NAK because either way the ping is complete, but actually I don't think that's correct; if the PING response is a NAK then the host is supposed to PING again before it can proceed to send OUT. So actually I think we should only accept ACK here.